### PR TITLE
Prevent deletion of commit if referred to by branch or tag

### PR DESCRIPTION
**Pull request by anandjakhaniya**

### CODE REVIEW

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java [hunk 1]
```diff
@@ -207,6 +211,37 @@ public CommitEntity getCommitEntity(
...
+              .append(" br where br.id.repository_id = :repositoryId ")
+              .append(" AND br.commit_hash = :commitHash ");
+      Query getBranchByCommitQuery = session.createQuery(getBranchByCommitHQLBuilder.toString());
+      getBranchByCommitQuery.setParameter("repositoryId", repositoryEntity.getId());
+      getBranchByCommitQuery.setParameter("commitHash", commitEntity.getCommit_hash());
+      BranchEntity branchEntity = (BranchEntity) getBranchByCommitQuery.uniqueResult();
+      if (branchEntity != null) {
+        throw new ModelDBException(
+            "Commit is associated with branch name : " + branchEntity.getId().getBranch(),
+            Code.INVALID_ARGUMENT);
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java]: Code.FAILED_PRECONDITION

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java [hunk 2]
```diff
@@ -207,6 +211,37 @@ public CommitEntity getCommitEntity(
...
+              .append(" AND lm.id.")
+              .append(ModelDBConstants.ENTITY_TYPE)
+              .append(" = :entityType")
+              .toString();
+      Query query = session.createQuery(getLabelsHql);
+      query.setParameter("entityHash", commitEntity.getCommit_hash());
+      query.setParameter("entityType", IDTypeEnum.IDType.VERSIONING_COMMIT_VALUE);
+      List<LabelsMappingEntity> labelsMappingEntities = query.list();
+      if (labelsMappingEntities.size() > 0) {
+        throw new ModelDBException("Commit is associated with Label", Code.INVALID_ARGUMENT);
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java]: Code . FAILED_PRECONDITION

File: backend/src/test/java/ai/verta/modeldb/BranchTest.java [hunk 3]
```diff
@@ -693,15 +713,23 @@ public void branchTest() {
...
-    Collections.reverse(commitShaList);
     commitShaList.forEach(
         commitSha -> {
           DeleteCommitRequest deleteCommitRequest =
               DeleteCommitRequest.newBuilder()
                   .setRepositoryId(RepositoryIdentification.newBuilder().setRepoId(id).build())
                   .setCommitSha(commitSha)
                   .build();
-          versioningServiceBlockingStub.deleteCommit(deleteCommitRequest);
+          if (commitShaList.get(3).equals(commitSha)) {
```
Comment (ravishetye) [inline backend/src/test/java/ai/verta/modeldb/BranchTest.java:718]: add either inline comment or a block comment for the function on what condition you are testing. Nothing fancy, even the current PR title can work

**Review by ravishetye — CHANGES_REQUESTED** [review]

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java (hunk 1 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java]: Changed

File: backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java (hunk 2 above)
Comment (anandjakhaniya) [reply backend/src/main/java/ai/verta/modeldb/versioning/CommitDAORdbImpl.java]: Changed

File: backend/src/test/java/ai/verta/modeldb/BranchTest.java (hunk 3 above)
Comment (anandjakhaniya) [reply backend/src/test/java/ai/verta/modeldb/BranchTest.java:718]: Added

**Pull request merged**